packages/client/lib/sync/skeleton.ts

```diff
@@ -252,17 +249,85 @@ export class Skeleton extends MetaDBManager {
+      this.config.logger.debug(
+        `Beacon chain extended new head=${lastchain.head} tail=${lastchain.tail} next=${short(
+          lastchain.next
+        )}`
+      )
+    }
+    return false
+  }
 
-    // Update the database with the new sync stats and insert the new
-    // head header. We won't delete any trimmed skeleton headers since
-    // those will be outside the index space of the many subchains and
-    // the database space will be reclaimed eventually when processing
-    // blocks above the current head.
-    await this.putBlock(head)
-    lastchain.head = number
-    await this.writeSyncStatus()
+  /**
+   * Announce and integrate a new head.
+   * @returns if the new head causes a reorg.
+   */
+  async setHead(head: Block, force = true, init = false, reorgthrow = false): Promise<boolean> {
+    return this.runWithLock<boolean>(async () => {
+      this.config.logger.debug(
+        `New skeleton head announced number=${head.header.number} hash=${short(
+          head.hash()
+        )} force=${force}`
+      )
+      const reorg = await this.processNewHead(head, force)
+      if (force && reorg) {
+        // It could just be a reorg at his head with previous tail preserved
+        const subchain = this.status.progress.subchains[0]
+        const parent = await this.getBlock(head.header.number - BigInt(1))
+        if (
+          subchain === undefined ||
+          parent === undefined ||
+          !parent.hash().equals(head.header.parentHash)
+        ) {
+          const s = {
+            head: head.header.number,
+            tail: head.header.number,
+            next: head.header.parentHash,
+          }
+          this.status.progress.subchains.unshift(s)
+          this.config.logger.info(
+            `Created new subchain head=${s.head} tail=${s.tail} next=${short(s.next)}`
+          )
+        } else {
+          // Only the head differed, tail is preserved
+          subchain.head = head.header.number
+        }
+      }
+      // Put this block irrespective of the force
+      await this.putBlock(head)
+      if (init) {
+        await this.trySubChainsMerge()
+      }
+      if ((force && reorg) || init) {
+        this.linked = await this.checkLinked()
+      }
+      if (force || init) {
+        await this.writeSyncStatus()
+      }
+      if (force && this.linked) {
+        void this.fillCanonicalChain()
+      }
+      // Earlier we were throwing on reorg, essentially for the puposes for killing the reverse fetcher
+      // but it can be handled properly in the calling fn without erroring
+      if (reorg && reorgthrow) {
+        if (force) {
+          throw errSyncReorged
+        } else {
+          throw errReorgDenied
+        }
+      }
+      return reorg
+    })
+  }
 
-    return false
+  /**
+   * Setup the skeleton to init sync with head
+   * @returns if the skeleton was reorged trying to init
+   */
+  async initSync(head: Block, reorgthrow = false): Promise<boolean> {
```

Comment:
Can you define the parameters and also explicitly state what the return value should be?

Reply:
done